sdk/scheduler/src/main/java/org/apache/mesos/offer/constrain/MaxPerHostnameRule.java

```diff
@@ -0,0 +1,138 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * Ensures that no more than N instances of a task type are running on a given hostname (or hostname
+ * pattern).
+ *
+ * For example, this can ensure that no more than N tasks are running against the 'rack:foo'
+ * attribute (exact match), or it can ensure that no distinct 'rack:.*' value has more than N tasks
+ * running against it (wildcarded grouping).
+ *
+ * To illustrate, let's look at a deployment scenario of 3 hosts with the following tasks:
+ *
+ *  hostname |     tasks
+ * ----------+---------------
+ *   host-1  | a-1, b-1, c-1
+ *   host-2  | a-2, c-2, c-3
+ *   host-3  | b-2, c-4
+ *
+ * In this example, let's assume a {@link MaxPerHostnameRule} with a limit of 2 and a regex of
+ * 'c-.*'. In this case, we're ensuring that no more than 2 tasks starting with 'c-' can be launched
+ * on any given host. Among the above three hosts, any offers from host-1 and host-3 will be
+ * accepted and offers from host-2 will be denied.
+ */
+public class MaxPerHostnameRule implements PlacementRule {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MaxPerHostnameRule.class);
+
+    private final int maxTasksPerHostname;
+    private final StringMatcher taskFilter;
+
+    /**
+     * Creates a new rule which will block deployment on tasks which already have N instances
+     * running against a specified attribute, with no filtering on task names (all tasks across the
+     * service are counted against the max).
+     *
+     * @param maxTasksPerSelectedHostname the maximum number of tasks which may be run on a selected
+     *     agent
+     */
+    public MaxPerHostnameRule(int maxTasksPerSelectedHostname) {
+        this(maxTasksPerSelectedHostname, null);
+    }
+
+    /**
+     * Creates a new rule which will block deployment on tasks which already have N instances
+     * running against a specified hostname.
+     *
+     * @param maxTasksPerSelectedHostname the maximum number of tasks which may be run on a selected
+     *     agent
+     * @param taskFilter a filter on task names to determine which tasks are included in the count,
+     *     for example counting all tasks, or just counting tasks of a given type
+     */
+    @JsonCreator
+    public MaxPerHostnameRule(
+            @JsonProperty("max") int maxTasksPerSelectedHostname,
+            @JsonProperty("task_filter") StringMatcher taskFilter) {
+        this.maxTasksPerHostname = maxTasksPerSelectedHostname;
+        if (taskFilter == null) { // null when unspecified in serialized data
+            taskFilter = AnyMatcher.create();
+        }
+        this.taskFilter = taskFilter;
+    }
+
+    @Override
+    public Offer filter(Offer offer, OfferRequirement offerRequirement, Collection<TaskInfo> tasks) {
+        int offerHostnameTaskCounts = 0;
```

Comment:
I don't quite have a full picture of the logic in my head here -- this rule doesn't keep any state about tasks that have already been scheduled, so is `tasks` a comprehensive list of all tasks that have already been placed as well as those that need to be placed? How is the constraint maintained?